File: 0212-Word-Search-II.py
class TrieNode:
    def __init__(self):
        self.children = {}
        self.isWord = False
        
    def addWord(self, word):
        cur = self
        for c in word:
            if c not in cur.children:
                cur.children[c] = TrieNode()
            cur = cur.children[c]
        cur.isWord = True

        
class Solution:
    def findWords(self, board: List[List[str]], words: List[str]) -> List[str]:
        root = TrieNode()
        for w in words:
            root.addWord(w)
            
        ROWS, COLS = len(board), len(board[0])
        res, visit = set(), set()
        
        def dfs(r, c, node, word):
            if (r < 0 or c < 0 or 
                r == ROWS or c == COLS or
                board[r][c] not in node.children or (r, c) in visit):
                return
            
            visit.add((r, c))
            node = node.children[board[r][c]]
            word += board[r][c]
            if node.isWord:
                res.add(word)
            
            dfs(r + 1, c, node, word)
            dfs(r - 1, c, node, word)
            dfs(r, c + 1, node, word)
            dfs(r, c - 1, node, word)
            visit.remove((r, c))
        
        for r in range(ROWS):
            for c in range(COLS):
                dfs(r, c, root, "")
        
        return list(res)

File: swift/0212-word-search-ii.swift
/**
 * Question Link: https://leetcode.com/problems/word-search-ii/
 */

class Solution {
    class TrieNode {
        var children =  [Character: TrieNode]()
        var word: String = ""
    }

    func findWords(_ board: [[Character]], _ words: [String]) -> [String] {
        var answer = [String]()
        let root = buildTrie(words)
    
        var board = board
    
        for row in 0..<board.count {
            for col in 0..<board[0].count {
                if root.children[board[row][col]] != nil {
                    dfs(row, col, root,&board,&answer)
                }
            }
        }

        return answer
    }

    func buildTrie(_ words: [String]) -> TrieNode {
        let root = TrieNode()
        for word in words {
            var current = root

            word.forEach {
                if current.children[$0] == nil {
                    current.children[$0] = TrieNode()
                }
                current = current.children[$0]!
            }
            current.word = word
        }

         return root
    }

     func getInBoardChild(_ x: Int, _ y: Int,_ board:[[Character]],_ root: TrieNode) -> TrieNode? {
        if 0..<board[0].count ~= x && 0..<board.count ~= y && board[y][x] != "." {
            return root.children[board[y][x]]
        }
        return nil
    }
    
    func dfs( _ y: Int,_ x: Int,_ root: TrieNode,_ board: inout [[Character]],_ answer: inout [String]) {
        guard let child = getInBoardChild(x, y, board, root) else { return }

        let char = board[y][x]
        board[y][x] = "."

        if !child.word.isEmpty {
            answer.append(child.word)
            child.word = ""
        }

        for d in [(x-1,y),(x+1,y),(x,y-1),(x,y+1)] {
            dfs(d.1,d.0, child,&board,&answer)
        }

         board[y][x] = char
    }
}
